### PR TITLE
create github-actions-demo.

### PR DESCRIPTION
**Pull request by BereznaiOleksandra**

### CODE REVIEW

**Review by kutsevol — APPROVED** [review]

File: .github/workflow/github-actions-demo.yml
Comment (kutsevol) [inline .github/workflow/github-actions-demo.yml:1]: Please update this workflow file with the same step as in pre-config-yaml file

File: app/main.py [hunk 1]
```diff
@@ -1,13 +1,39 @@
 import datetime
+from abc import ABC, abstractmethod
 from dataclasses import dataclass
 
 
+class BookDiscount(ABC):
```
Comment (kutsevol) [inline app/main.py]: It's better to name this class with `Abstract*` prefix for clear understanding. I think there are several possible names: `AbstractDiscount` or `AbstractPriceDiscount`. Why not `BookDiscount`, because in my opinion discount is more related to price than for book :) And if you have some class in the future, for example `Magazine` with the same discount strategies, the names will be unequal.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    hooks:
+      - id: ruff
+        args: [ --fix ]
+
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v1.10.0
+    hooks:
+      - id: mypy
+
+  - repo: https://github.com/pre-commit/pre-commit-hooks
```
Comment (kutsevol) [inline .pre-commit-config.yaml]: Please remove duplicate pre-commit-hooks

File: app/main.py [hunk 3]
```diff
@@ -70,6 +98,24 @@ def get_book_list(self) -> list[Book]:
...
         self.book_list = book_list
 
-
     def add_book(self, book: Book):
         self.book_list.append(book)
+
+
+class LibraryCatalog:
+    def __init__(self):
+        self.books = []
```
Comment (kutsevol) [inline app/main.py]: I think this is not the best data structure for storing a catalog, what else do you think would be better?

**Review by kutsevol — CHANGES_REQUESTED** [review]

File: app/main.py (hunk 3 above)
Comment (BereznaiOleksandra) [reply app/main.py]: self.books: List[Book] = []

Like this?

Comment (BereznaiOleksandra) [reply app/main.py]: or dict?

Comment (kutsevol) [reply app/main.py]: Yeah! Dict is better :) Please use explicit typing for dict -> `dict[key_type, value_type]`

File: tests/test_main.py [hunk 4]
```diff
@@ -22,9 +24,25 @@ def test_book_older_than_5_years():
...
+
+
+def test_book_fix_discount():
+    book = BookFactory.build(price=200, discount=BookFixDiscount(20))
+    assert book.final_price() == 180
+
+
+def test_book_persent_discount():
+    book = BookFactory.build(price=200, discount=BookPersentDiscount(30))
+    assert book.final_price() == 140
```
Comment (kutsevol) [inline tests/test_main.py:48]: Please add a test where we initialize a book instance with the default (zero) discount strategy, then sequentially change its discount strategy to different ones (e.g., fixed, percent). For each change, check that the final price updates accordingly and that discounts do not accumulate from previous strategies—only the current strategy should be applied.

Example pseudocode:

```python
book = BookFactory.build(price=100)
assert book.final_price() == 100
# without creating new instance of class please use this strategy BookFixDiscount(20)
assert book.final_price() == 80

# the same logic like previous line BookPersentDiscount(30)
assert book.final_price() == 70

# Make sure the discount is not cumulative:
# The final price should be 70 (30% off 100), not 80 - 30% or any other combination.
```

**Review by kutsevol — APPROVED** [review]
